lib/model/recent_senders.dart

```diff
@@ -0,0 +1,167 @@
+import 'package:collection/collection.dart';
+import 'package:flutter/foundation.dart';
+
+import '../api/model/events.dart';
+import '../api/model/model.dart';
+
+/// A data structure to keep track of stream and topic messages of users (senders).
+///
+/// Use [latestMessageIdOfSenderInStream] and [latestMessageIdOfSenderInTopic]
+/// to get the relevant data.
+class RecentSenders {
+  // streamSenders[streamId][senderId] = MessageIdTracker
+  @visibleForTesting
+  final Map<int, Map<int, MessageIdTracker>> streamSenders = {};
+
+  // topicSenders[streamId][topic][senderId] = MessageIdTracker
+  @visibleForTesting
+  final Map<int, Map<String, Map<int, MessageIdTracker>>> topicSenders = {};
+
+  int? latestMessageIdOfSenderInStream({
+    required int streamId,
+    required int senderId,
+  }) => streamSenders[streamId]?[senderId]?.maxId;
+
+  int? latestMessageIdOfSenderInTopic({
+    required int streamId,
+    required String topic,
+    required int senderId,
+  }) => topicSenders[streamId]?[topic]?[senderId]?.maxId;
+
+  /// Calls [handleMessage] for each of the message.
+  ///
+  /// [inReverse] is used for the optimization of the underlying
+  /// [MessageIdTracker.add] method. It should be `false` (default) when
+  /// [messages] are the initial messages in a narrow (fetched through
+  /// [MessageListView.fetchInitial]), and `true` when they are the older
+  /// messages in a narrow (fetched through [MessageListView.fetchOlder]).
+  void handleMessages(List<Message> messages, {bool inReverse = false}) {
+    for (int i = inReverse ? messages.length - 1 : 0;
+                 inReverse ? i >= 0              : i < messages.length;
+                 inReverse ? i--                 : i++) {
+      handleMessage(messages[i]);
+    }
```

Comment:
In the case of `MessageListView.fetchOlder`, by traversing the list of messages in reverse order, message IDs go into the start or the end of the tracker list. But in the normal order, they'll also go in the middle of the tracker list. Now that we use `QueueList` (which takes $O(1)$ for both `addFirst` and `addLast`) as the type of ID tracker list, traversing in reverse will result in optimization.

Reply:
I am having a slightly hard time coming up with an efficient solution for this. 😃 
Here's what I have thought of the `handleMessages` method so far, especially, determining all the affected `MessageIdTracker`s:
```dart
    final streamTracker = <(int, int), List<int>>{};
    final topicTracker = <(int, String, int), List<int>>{};

    for (final message in messages) {
      if (message is! StreamMessage) return;
      final StreamMessage(:streamId, :topic, :senderId, id: int messageId) = message;
      (streamTracker[(streamId, senderId)] ??= []).add(messageId);
      (topicTracker[(streamId, topic, senderId)] ??= []).add(messageId);
    }

    for (final entry in streamTracker.entries) {
      final (streamId, senderId) = entry.key;
      final tracker = (streamSenders[streamId] ??= {})[senderId] ??= MessageIdTracker();
      tracker.addAll(entry.value);
    }
    for (final entry in topicTracker.entries) {
      final (streamId, topic, senderId) = entry.key;
      final tracker = ((topicSenders[streamId] ??= {})[topic] ??= {})[senderId] ??= MessageIdTracker();
      tracker.addAll(entry.value);
    }
```
But as you can see, it already has 3 for loops. Then when I want to have a single method for adding the relevant message IDs to a `MessageIdTracker`, there's the need to check that we don't add the already-existing IDs from the incoming `id`s list. So I think it again needs to call `add` method for each message.

Reply:
Having 3 loops is fine - they're happening in sequence, not nested, so the total cost is only the sum of the three loops' respective cost. And for example three times $O(f(n))$ is still $O(f(n))$.

Then for the implementation of addAll, it doesn't need to be $O(n+m)$ (for $n$ new messages and $m$ existing messages) — the goal is only for it to be $O((n+m) \log (n+m))$. In the implementation of `add`, the expensive part is the actual insertion, moving elements around, which can take $m$ steps. The search is a binary search, which is much cheaper at $O(\log m)$ steps. So addAll can do $n$ searches just fine — the thing it has to avoid is $n$ individual insertions.

Reply:
Here's the revised implementation for `addAll`:
```dart
  void addAll(List<int> ids) {
    final idsToBeAdded = <int>[];

    if (_ids.isEmpty) {
      _ids.addAll(ids);
      return;
    }

    for (final id in ids) {
      final i = lowerBound(_ids, id);
      if (i < _ids.length && _ids[i] == id) {
        // The ID is already present. Nothing to do.
        continue;
      }
      idsToBeAdded.add(id);
    }

    _ids.insertAll(0, idsToBeAdded);
  }
```
Am I on the right track?

Reply:
Yeah, that's a good direction. What's a bound (in big-O notation) on how long that will take?

Then after that insertAll, the IDs in the tracker may be out of order. So that will need to be fixed.

Reply:
For `addAll`, I think it takes $O(nlogm) + O(n + m)$ which is $O(nlogm)$ at the end of the day, if I am not wrong.

But thinking about the current implementation in this PR where there's `inReverse` in `handleMessages`, I think it takes less time possible, just $O(n)$. For each message id, it just adds it to the start or the end of a `MessageIdTracker` list (thanks to `inReverse`). Now that we use `QueueList` (`Deque`), it just takes $O(1)$ and eventually making the whole adding operation take $O(n)$!

Reply:
> But thinking about the current implementation in this PR where there's `inReverse` in `handleMessages`, I think it takes less time possible, just O(n). For each message id, it just adds it to the start or the end of a `MessageIdTracker` list (thanks to `inReverse`). Now that we use `QueueList` (`Deque`), it just takes O(1) and eventually making the whole adding operation take O(n)!

The thing is that the messages might not go at the very start or end. For example, consider what happens in a stream's tracker if the user first visits a topic narrow within that stream (and maybe scrolls back through a bunch of history), then visits the stream narrow itself.

We're also getting close to properly supporting moved messages, #150 — see #787. So after both this PR and that one are merged (or perhaps as part of a later revision of #787 after this PR is merged), we'll be handling message moves on this RecentSenders data structure too. Then possible scenarios include:
* Some older messages get moved into a topic. They go before lots of more recent history in the topic; but our client hasn't yet loaded any of that history. The topic ID tracker ends up with just those older messages.

  Then later the user visits that topic narrow. The first fetched messages are after, not before, all the existing messages. They scroll up further… and the messages from `fetchOlder` go in between a bunch of older messages and a bunch of newer messages.

* Some messages get moved into a topic, after the user has already scrolled through a lot of history in the topic so we have a long list of messages there. The move won't go through this `handleMessages` code path exactly, so that's a problem we could wait to solve until we implement moves; but we'll need to solve it then, and the same solution that handles the other scenarios will handle this one.

Reply:
> For `addAll`, I think it takes O(nlogm)+O(n+m) which is O(nlogm) at the end of the day, if I am not wrong.

Yep, the `addAll` you quoted above takes time bounded by $O(n \log m) + O(n + m)$ ­— the first term for the loop that calls `lowerBound`, then the second for the `insertAll` call.

That can be simplified to
$O(n \log m + n + m)$;
and then to
$O(n \log m + m)$
because $n$ is bounded by $n \log m$.

It's not true that it's $O(n \log m)$, though. Either $m$ could be bigger than $n \log m$, or vice versa, so neither term can be simplified away.

But the goal I mentioned above was a bound of $O((n + m) \log (n + m))$, so it's still under that bound and we're still in good shape.

The `addAll` above still needs another bit of logic at the end to deal with the fact that `insertAll` may leave the IDs in the tracker out of order, though.

(Formatting tip: write `\log`, with a backslash, to get it formatted as the log operator, with an upright font and spacing around it.)

Reply:
Thanks for the explanation, it was really helpful; also thanks for the formation tip!

Here's the new revision for `addAll`, still with $O(n\log m + m)$:

```dart
  void addAll(List<int> ids) {
    if (_ids.isEmpty) {
      _ids.addAll(ids);
      return;
    }

    final idsToBeAdded = <int>[];
    for (final id in ids) {
      final i = lowerBound(_ids, id);
      if (i < _ids.length && _ids[i] == id) {
        // The ID is already present. Nothing to do.
        continue;
      }
      idsToBeAdded.add(id);
    }

    _ids = setUnion(_ids, idsToBeAdded);
  }
```

Reply:
Great. Looking forward to seeing the revision!